File: svc-data/src/test/java/com/beancounter/marketdata/FxRateTests.java
package com.beancounter.marketdata;

import static org.assertj.core.api.Assertions.assertThat;

import com.beancounter.common.contracts.FxPairResults;
import com.beancounter.common.model.Currency;
import com.beancounter.common.model.FxRate;
import com.beancounter.common.model.IsoCurrencyPair;
import com.beancounter.common.utils.RateCalculator;
import com.beancounter.marketdata.providers.fxrates.EcbRates;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.File;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.springframework.core.io.ClassPathResource;

class FxRateTests {
  private static IsoCurrencyPair USD_USD = IsoCurrencyPair.builder().from("USD").to("USD").build();
  private static IsoCurrencyPair AUD_NZD = IsoCurrencyPair.builder().from("AUD").to("NZD").build();
  private static IsoCurrencyPair NZD_AUD = IsoCurrencyPair.builder().from("NZD").to("AUD").build();
  private static IsoCurrencyPair AUD_USD = IsoCurrencyPair.builder().from("AUD").to("USD").build();
  private static IsoCurrencyPair USD_AUD = IsoCurrencyPair.builder().from("USD").to("AUD").build();

  @Test
  void is_FxRateResponseSerializing() throws Exception {
    File jsonFile = new ClassPathResource("contracts/ecb/ecbEarly.json").getFile();
    EcbRates ecbRates = new ObjectMapper().readValue(jsonFile, EcbRates.class);
    assertThat(ecbRates)
        .isNotNull()
        .hasNoNullFieldsOrProperties()
    ;
    assertThat(ecbRates.getRates()).hasSize(6);
  }

  @Test
  void is_RateCalculatorComputing() {

    Collection<IsoCurrencyPair> pairs =
        getCurrencyPairs(USD_USD, AUD_NZD, NZD_AUD, AUD_USD, USD_AUD);

    Map<String, FxRate> rates = getRateTable();
    String asAt = "2019-11-21";

    FxPairResults values = RateCalculator.compute(asAt, pairs, rates);

    FxRate audUsd = values.getRates().get(AUD_USD);
    FxRate usdAud = values.getRates().get(USD_AUD);
    // Verify that the inverse rate is equal
    BigDecimal calc = BigDecimal.ONE.divide(audUsd.getRate(), 8, RoundingMode.HALF_UP);
    assertThat(usdAud.getRate()).isEqualTo(calc);

  }

  private Collection<IsoCurrencyPair> getCurrencyPairs(IsoCurrencyPair... pairs) {
    Collection<IsoCurrencyPair> results = new ArrayList<>();
    Collections.addAll(results, pairs);
    return results;
  }

  private Map<String, FxRate> getRateTable() {
    Map<String, FxRate> rates = new HashMap<>();
    rates.put("NZD", getRate("NZD", "1.5536294691"));
    rates.put("AUD", getRate("AUD", "1.48261"));
    rates.put("USD", getRate("USD", "1"));
    return rates;
  }

  private FxRate getRate(String to, String rate) {
    return FxRate.builder()
        .rate(new BigDecimal(rate))
        .from(Currency.builder().code("USD").build())
        .to(
            Currency.builder().code(to).build())
        .build();
  }
}

File: svc-data/src/test/java/com/beancounter/marketdata/FxRateTests.kt
package com.beancounter.marketdata

import com.beancounter.common.model.Currency
import com.beancounter.common.model.FxRate
import com.beancounter.common.model.IsoCurrencyPair
import com.beancounter.common.utils.BcJson
import com.beancounter.common.utils.RateCalculator
import com.beancounter.marketdata.Constants.Companion.AUD
import com.beancounter.marketdata.Constants.Companion.NZD
import com.beancounter.marketdata.Constants.Companion.USD
import com.beancounter.marketdata.integ.FxMvcTests.Companion.fxMock
import com.beancounter.marketdata.providers.fxrates.EcbRates
import com.fasterxml.jackson.databind.ObjectMapper
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.springframework.core.io.ClassPathResource
import java.math.BigDecimal
import java.math.RoundingMode

/**
 * Unit test FX assumptions.
 */
internal class FxRateTests {
    private val objectMapper: ObjectMapper = BcJson().objectMapper
    @Test
    @Throws(Exception::class)
    fun is_FxRateResponseSerializing() {
        val jsonFile = ClassPathResource("$fxMock/ecbEarly.json").file
        val ecbRates = objectMapper.readValue(jsonFile, EcbRates::class.java)
        Assertions.assertThat(ecbRates)
            .isNotNull
            .hasNoNullFieldsOrProperties()
        Assertions.assertThat(ecbRates.rates).hasSize(6)
    }

    @Test
    fun is_RateCalculatorComputing() {
        val pairs = getCurrencyPairs(USD_USD, AUD_NZD, NZD_AUD, AUD_USD, USD_AUD)
        val rates = rateTable
        val asAt = "2019-11-21"
        val (rates1) = RateCalculator.compute(asAt, pairs, rates)
        val audUsd = rates1[AUD_USD]
        val usdAud = rates1[USD_AUD]
        // Verify that the inverse rate is equal
        val calc = BigDecimal.ONE.divide(audUsd!!.rate, 8, RoundingMode.HALF_UP)
        Assertions.assertThat(usdAud!!.rate).isEqualTo(calc)
    }

    private fun getCurrencyPairs(vararg pairs: IsoCurrencyPair): Collection<IsoCurrencyPair> {
        val results = ArrayList<IsoCurrencyPair>()
        for (pair in pairs) {
            results.add(pair)
        }
        return results
    }

    private val rateTable: Map<String, FxRate>
        get() {
            val rates: MutableMap<String, FxRate> = HashMap()
            rates[NZD.code] = getRate(NZD.code, "1.5536294691")
            rates[AUD.code] = getRate(AUD.code, "1.48261")
            rates[USD.code] = getRate(USD.code, "1")
            return rates
        }

    private fun getRate(to: String, rate: String): FxRate {
        return FxRate(USD, Currency(to), BigDecimal(rate), null)
    }

    companion object {
        private val USD_USD = IsoCurrencyPair(USD.code, USD.code)
        private val AUD_NZD = IsoCurrencyPair(AUD.code, NZD.code)
        private val NZD_AUD = IsoCurrencyPair(NZD.code, AUD.code)
        private val AUD_USD = IsoCurrencyPair(AUD.code, USD.code)
        private val USD_AUD = IsoCurrencyPair(USD.code, AUD.code)
    }
}
